### DISCUSSION TITLE
Client-side exception and custom error page

### DISCUSSION THREAD

**Discussion by ale-grosselle** [vercel/next.js#36469] [Help] [answered]
When a client-side exception occurred and the custom error page is rendered, we want to send from the server side a log to our internal system.

I don't know how to understand when the error page is rendered using "client rendering" implementation.

I need to implement "  if(error is from client) {" part of the code

```ts
function ErrorComponent({ statusCode }: { statusCode: number }) {
 
  return <Error {...template} />;
}

ErrorComponent.getInitialProps = ({
  res,
  err,
  req,
  ...others
}: NextPageContext) => {
  //To implement
  if(error is from client) {
    log to kibana
  }
  return { statusCode };
};

export default ErrorComponent;


```

**Comment by icyJoseph** [comment, marked as answer]
Hi,

I am a bit confused here. On the client you can use `ErrorBoundaries` and the `getDerivedStateFromError` and `componentDidCatch` methods for reporting errors when rendering a page.

If you are talking about these: https://nextjs.org/docs/advanced-features/custom-error-page#more-advanced-error-page-customizing, then **I think**, you can do something like:

```ts
Error.getInitialProps = ({ res, err }) => {
  const isServer = Boolean(res)
  if(isServer){
    // do something
  } else {
    // do something else
  }
  // proceed
  const statusCode = isServer ? res.statusCode : err ? err.statusCode : 404
  return { statusCode }
}
```

I base the `isServer` check validity from, https://nextjs.org/docs/api-reference/data-fetching/get-initial-props#context-object, where the context object for `getInitialProps` contains `req`/`res` only on the server. So the lack of `req`/`res` would tell us that we are on the client.

Then again, I did not follow so well your question, so I might be way off here.

**Comment by ale-grosselle** [comment]
@icyJoseph yes, I'm talking about the error custom page.

The error's page `getInitialProps` method is always runs on the server side.

For example, when I click 5 times the counter, the error page is rendered:
![ezgif-1-8b26921bba](https://user-images.githubusercontent.com/6173088/165258683-c4551f95-d617-4150-b6a6-3d0ccce917f5.gif)

The buggyCounter:

```ts
import React from 'react';

export const BuggyCounter = function () {
  const [counter, setCounter] = React.useState(0);

  const handleClick = () => {
    setCounter(counter + 1);
  };

  if (counter === 5) {
    // Simulate a JS error
    throw new Error('I crashed!');
  }
  return <button onClick={handleClick}>{counter}</button>;
};

```

```ts
<div className={styles.main}>
      <BuggyCounter />
</div>
```

The UI of custom error page is rendered when the counter is equal to 5 and the `Error.getInitialProps` always runned on the server.

I want to know on `Error.getInitialProps` when is triggered by a client side error:

```ts
ErrorComponent.getInitialProps = ({
  res,
  err,
  req,
  ...others
}: NextPageContext) => {
  //To implement
  if(error is from a client side exception) {
    log to kibana
  }
  return { statusCode };
};
```

**Reply by icyJoseph** [reply]
Hi, but are you absolutely sure? I just used your counter on `pages/index.tsx`, and defined `_error.tsx`, and see that the `getInitialProps` function is run on the client, done with `next build` and `next start`.

```tsx
import type { NextPage } from "next";

const Error: NextPage<{ statusCode?: number }> = ({ statusCode }) => {
  console.log({ statusCode });
  return (
    <p>
      {statusCode
        ? `An error ${statusCode} occurred on server`
        : "An error occurred on client"}
    </p>
  );
};

// various log checks
Error.getInitialProps = ({ res, err }) => {
  console.log("err", err);
  console.log("res", res);
  const clientSideError = res ? res.statusCode : Boolean(err);
  console.log({ clientSideError });
  console.log("server", Boolean(res));
  const statusCode = res ? res.statusCode : err ? err.statusCode : 404;
  return { statusCode };
};

export default Error;
```

<img width="1507" alt="Screenshot 2022-04-26 at 11 02 46" src="https://user-images.githubusercontent.com/21013447/165264004-5d5c324b-790f-4af1-ab50-1548dd5fbb9d.png">

**Comment by ale-grosselle** [comment]
@icyJoseph You agree it runs on clientside and I can follow your solution!

Many thanks

**Reply by icyJoseph** [reply]
Awesome, don't forget to make it as answered, I think I'd like to come back to this later on.
